### PR TITLE
Support configure static IP 

### PR DESCRIPTION
**Pull request by futuretea**
**DHCP**
image
image

image

**Static**
image
image

**Related Issue:**
rancher/harvester#367
rancher/harvester#470
rancher/harvester#471
rancher/harvester#490

### CODE REVIEW

File: pkg/console/install_panels.go [hunk 1]
```diff
@@ -498,6 +546,350 @@ func addNetworkPanel(c *Console) error {
...
+				return err.Error(), nil
+			}
+		}
+		installData.DNSServers = dnsServers
+		cfg.Config.K3OS.DNSNameservers = dnsServerList
+
+		// run-cmd
+		configureNetworkManual := fmt.Sprintf("/sbin/configure-network-manual %s %s %s %s %s",
+			installData.Interface, installData.IP, installData.NetMask, installData.Gateway, strings.Join(dnsServerList, " "))
+		cfg.Config.Runcmd = []string{configureNetworkManual}
```
Comment (bk201) [inline pkg/console/install_panels.go:853]: Does the static IP take effect only after installation?
If that's the case, the fetching URL behavior in SSH keys and Cloud Config dialog might not work.

Comment (futuretea) [reply pkg/console/install_panels.go:853]: During installation，live os will use dhcp to fetch URL.

Comment (gitlawr) [reply pkg/console/install_panels.go:793]: Some users request to install with no DHCP. Sounds like this does not satisfy their requirements

**Review by bk201 — APPROVED** [review]
LGTM. Thanks for adapting the config structure change.

File: pkg/console/util.go [hunk 2]
```diff
@@ -90,14 +90,13 @@ func getRemoteSSHKeys(url string) ([]string, error) {
...
 }
 
-func getFormattedServerURL(addr string) string {
-	if !strings.HasPrefix(addr, "https://") {
-		addr = "https://" + addr
+func getFormattedServerURL(addr string) (string, error) {
+	ipErr := checkIP(addr)
+	domainErr := checkDomain(addr)
+	if ipErr != nil && domainErr != nil {
+		return "", fmt.Errorf("%s is not an valid ip/domain", addr)
```
Comment (gitlawr) [inline pkg/console/util.go:97]: s/an/a/

File: pkg/console/console.go [hunk 3]
```diff
@@ -92,6 +93,23 @@ func (c *Console) setContentByName(name string, content string) error {
 	return err
 }
 
+func (c *Console) CloseElement(name string) error {
```
Comment (gitlawr) [inline pkg/console/console.go:96]: Seems like we seldom handle the error in the caller. Can we log the error and do no error return for the function?

File: pkg/console/install_panels.go [hunk 4]
```diff
@@ -483,33 +510,313 @@ func addTokenPanel(c *Console) error {
...
+	dnsServersV, err := widgets.NewInput(c.Gui, dnsServersPanel, dnsServersLabel, false)
+	if err != nil {
+		return err
+	}
+
+	networkValidatorV := widgets.NewPanel(c.Gui, networkValidatorPanel)
+
+	gotoPrevPanel := func(c *Console, names ...string) func(g *gocui.Gui, v *gocui.View) error {
+		return func(g *gocui.Gui, v *gocui.View) error {
+			_ = c.CloseElement(networkValidatorPanel)
```
[the remark this answers is not in the document]
Comment (futuretea) [reply pkg/console/install_panels.go:572]: Indicates that this error was deliberately ignored, not forgotten to deal with.

File: pkg/console/console.go (hunk 3 above)
Comment (futuretea) [reply pkg/console/console.go:96]: yes

**Comment by gitlawr** [comment]
LGTM, can you squash the commits?

**Comment by futuretea** [comment]
> LGTM, can you squash the commits?

Updated

**Review by gitlawr — APPROVED** [review]

**Pull request merged**